sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala

```diff
@@ -965,12 +969,34 @@ class RocksDB(
 
   /**
    * Function to release RocksDB instance lock that allows for synchronized access to the state
-   * store instance
+   * store instance. Optionally provide a thread to check against, and release only if provided
+   * thread is the one that acquired the lock.
    *
    * @param opType - operation type releasing the lock
+   * @param forThreadOpt - optional thread to check against acquired thread
    */
-  private def release(opType: RocksDBOpType): Unit = acquireLock.synchronized {
+  private def release(opType: RocksDBOpType,
+    forThreadOpt: Option[AcquiredThreadInfo]): Unit = acquireLock.synchronized {
     if (acquiredThreadInfo != null) {
+      if (forThreadOpt.nonEmpty) {
+        if (forThreadOpt.get.threadRef.get.isEmpty) {
+          logInfo(log"Thread reference is empty when attempting to release for" +
+            log" opType=${MDC(LogKeys.OP_TYPE, opType.toString)}, ignoring release." +
+            log" Lock is held by ${MDC(LogKeys.THREAD, acquiredThreadInfo)}")
+          return
+        }
+
+        if (acquiredThreadInfo.threadRef.get.isDefined
+          && forThreadOpt.get.threadRef.get.isDefined
+          && acquiredThreadInfo.threadRef.get.get.getId
```

Comment:
can we not do object level equality comparison here ? or maybe add a comparator function if required ?